xain/grpc/participant.py

```diff
@@ -65,34 +118,240 @@ def end_training(
+            elif msg == coordinator_pb2.State.FINISHED:
+                st.state = ParState.DONE
+                st.cv.notify()
+        elif st.state == ParState.POST_TRAINING:
+            if msg == coordinator_pb2.State.STANDBY:
+                # not selected
+                st.state = ParState.WAITING_FOR_SELECTION
+                # prob ok to keep st.round as it is
+                st.cv.notify()
+            elif msg == coordinator_pb2.State.ROUND and r == st.round + 1:
+                st.state = ParState.TRAINING
+                st.round = r
+                st.cv.notify()
+            elif msg == coordinator_pb2.State.FINISHED:
+                st.state = ParState.DONE
+                st.cv.notify()
+
+
+def message_loop(chan, st, terminate):
+    """Periodically sends (and handles) heartbeat messages in a loop.
+
+    Args:
+        chan: gRPC channel to Coordinator.
+        st (obj:`StateRecord`): Participant state record.
+        terminate (obj:`threading.Event`): Event to terminate message loop.
+    """
+    coord = coordinator_pb2_grpc.CoordinatorStub(chan)
+    while not terminate.is_set():
+        req = coordinator_pb2.HeartbeatRequest()
+        reply = coord.Heartbeat(req)
+        transit(st, reply)
+        time.sleep(HEARTBEAT_TIME)
+
+
+def go(part):
+    """Top-level function for the Participant state machine.
+
+    After rendezvous and heartbeat initiation, the Participant is
+    WAITING_FOR_SELECTION. When selected, it moves to TRAINING followed by
+    POST_TRAINING. If selected again for the next round, it moves back to
+    TRAINING, otherwise it is back to WAITING_FOR_SELECTION.
+
+    Args:
+        part (obj:`Participant`): Participant object for training computation.
+    """
+    # use insecure channel for now
+    with grpc.insecure_channel("localhost:50051") as chan:  # thread-safe
+        rendezvous(chan)
+
+        st = StateRecord()
+        terminate = threading.Event()
+        ml = threading.Thread(target=message_loop, args=(chan, st, terminate))
+        ml.start()
+
+        # in WAITING_FOR_SELECTION state
+        begin_selection_wait(st, chan, part)
+
+        # possibly several training rounds later...
+        # in DONE state
+        terminate.set()
+        ml.join()
+
+
+def begin_selection_wait(st, chan, part):
+    """Perform actions in Participant state WAITING_FOR_SELECTION.
+
+    Args:
+        st (obj:`StateRecord`): Participant state record.
+        chan: gRPC channel to Coordinator.
+        part (obj:`Participant`): Participant object for training computation.
+    """
+    ps = st.wait_until_selected_or_done()
+    if ps == ParState.TRAINING:
+        # selected
+        begin_training(st, chan, part)
+    elif ps == ParState.DONE:
+        pass
 
-    try:
-        # never returns unless there is an exception
-        heartbeat_thread.join()
-    except KeyboardInterrupt:
-        terminate_event.set()
-        channel.close()
 
+def begin_training(st, chan, part):
```

Comment:
Also for future reference I would like to have a better way to organize this logic so that we don't have constant recursive calls between `begin_selection_wait` and `begin_training`

Reply:
The main reason for favouring function calls here was mainly for readability / close correspondence with the state transition system - e.g. you can think of a `begin_training` call to be a state transition to the `TRAINING` state. But that is maybe a matter of taste. I would be in favour of a different way if it is more readable or there is a performance impact.